<a href="https://colab.research.google.com/github/Sam-krish2411/DATA-SCIENCE-ASSIGNMENT/blob/main/NEURAL_NETWORKS.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

##**TASK-1:DATA EXPLORATION AND PREPROCESSING**

In [1]:
#Importing the libraries

import pandas as pd
import numpy as np
from sklearn.preprocessing import StandardScaler, LabelEncoder

In [2]:
data = pd.read_csv("Alphabets_data.csv")

In [3]:
print("First 5 rows of the dataset:\n", data.head())
print("\nDataset Info:\n")
print(data.info())
print("\nSummary Statistics:\n", data.describe())

First 5 rows of the dataset:
   letter  xbox  ybox  width  height  onpix  xbar  ybar  x2bar  y2bar  xybar  \
0      T     2     8      3       5      1     8    13      0      6      6   
1      I     5    12      3       7      2    10     5      5      4     13   
2      D     4    11      6       8      6    10     6      2      6     10   
3      N     7    11      6       6      3     5     9      4      6      4   
4      G     2     1      3       1      1     8     6      6      6      6   

   x2ybar  xy2bar  xedge  xedgey  yedge  yedgex  
0      10       8      0       8      0       8  
1       3       9      2       8      4      10  
2       3       7      3       7      3       9  
3       4      10      6      10      2       8  
4       5       9      1       7      5      10  

Dataset Info:

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 20000 entries, 0 to 19999
Data columns (total 17 columns):
 #   Column  Non-Null Count  Dtype 
---  ------  --------------  -----

In [4]:
# Check for missing values
print("\nMissing values in each column:\n", data.isnull().sum())



Missing values in each column:
 letter    0
xbox      0
ybox      0
width     0
height    0
onpix     0
xbar      0
ybar      0
x2bar     0
y2bar     0
xybar     0
x2ybar    0
xy2bar    0
xedge     0
xedgey    0
yedge     0
yedgex    0
dtype: int64


In [7]:
data.columns

Index(['letter', 'xbox', 'ybox', 'width', 'height', 'onpix', 'xbar', 'ybar',
       'x2bar', 'y2bar', 'xybar', 'x2ybar', 'xy2bar', 'xedge', 'xedgey',
       'yedge', 'yedgex'],
      dtype='object')

In [8]:
# Separate features and labels
X = data.drop("letter", axis=1)   # Features
y = data["letter"]                # Target

In [9]:
#  Encode labels (A, B, C… → 0, 1, 2…)
le = LabelEncoder()
y = le.fit_transform(y)

In [10]:
# Normalize features
scaler = StandardScaler()
X = scaler.fit_transform(X)

In [11]:
# Display dataset summary
print("\nNumber of samples:", X.shape[0])
print("Number of features:", X.shape[1])
print("Number of classes:", len(np.unique(y)))
print("Classes:", le.classes_)


Number of samples: 20000
Number of features: 16
Number of classes: 26
Classes: ['A' 'B' 'C' 'D' 'E' 'F' 'G' 'H' 'I' 'J' 'K' 'L' 'M' 'N' 'O' 'P' 'Q' 'R'
 'S' 'T' 'U' 'V' 'W' 'X' 'Y' 'Z']


##**TASK-2:MODEL IMPLEMENTATION**

In [12]:
#Importing the libraries

from sklearn.model_selection import train_test_split
from sklearn.metrics import accuracy_score, classification_report
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Dense

In [13]:
# Split data into training and test sets
X_train, X_test, y_train, y_test = train_test_split(
    X, y, test_size=0.2, random_state=42, stratify=y
)

In [14]:
#Construct a basic ANN model
model = Sequential()
model.add(Dense(64, input_dim=X_train.shape[1], activation='relu'))   # Hidden layer
model.add(Dense(len(np.unique(y)), activation='softmax'))             # Output layer


/usr/local/lib/python3.12/dist-packages/keras/src/layers/core/dense.py:93: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(activity_regularizer=activity_regularizer, **kwargs)


In [15]:
# Compile the model
model.compile(optimizer='adam', loss='sparse_categorical_crossentropy', metrics=['accuracy'])


In [16]:
# Train the model
history = model.fit(X_train, y_train, epochs=20, batch_size=32, validation_split=0.2, verbose=1)


Epoch 1/20
400/400 ━━━━━━━━━━━━━━━━━━━━ 2s 3ms/step - accuracy: 0.2891 - loss: 2.6446 - val_accuracy: 0.6347 - val_loss: 1.3917
Epoch 2/20
400/400 ━━━━━━━━━━━━━━━━━━━━ 2s 3ms/step - accuracy: 0.6719 - loss: 1.2528 - val_accuracy: 0.7450 - val_loss: 0.9861
Epoch 3/20
400/400 ━━━━━━━━━━━━━━━━━━━━ 1s 2ms/step - accuracy: 0.7396 - loss: 0.9534 - val_accuracy: 0.7784 - val_loss: 0.8260
Epoch 4/20
400/400 ━━━━━━━━━━━━━━━━━━━━ 1s 2ms/step - accuracy: 0.7907 - loss: 0.7897 - val_accuracy: 0.8081 - val_loss: 0.7289
Epoch 5/20
400/400 ━━━━━━━━━━━━━━━━━━━━ 2s 3ms/step - accuracy: 0.8057 - loss: 0.7055 - val_accuracy: 0.8209 - val_loss: 0.6658
Epoch 6/20
400/400 ━━━━━━━━━━━━━━━━━━━━ 2s 3ms/step - accuracy: 0.8190 - loss: 0.6329 - val_accuracy: 0.8353 - val_loss: 0.6147
Epoch 7/20
400/400 ━━━━━━━━━━━━━━━━━━━━ 1s 2ms/step - accuracy: 0.8307 - loss: 0.5952 - val_accuracy: 0.8422 - val_loss: 0.5781
Epoch 8/20
400/400 ━━━━━━━━━━━━━━━━━━━━ 1s 2ms/step - accuracy: 0.8457 - loss: 0.5350 - val_accuracy: 0.

####  Epoch-wise Performance Overview

- **Epoch 1**: Initial training accuracy was low (28.91%) with high loss (2.6446), but validation accuracy was already promising (63.47%), indicating potential for generalization.
- **Epochs 2–5**: Rapid improvement in both training and validation metrics. Accuracy rose to ~80%, and validation loss dropped below 0.7.
- **Epochs 6–10**: Continued steady gains. Training accuracy reached ~87%, validation accuracy crossed 86%, and loss values declined consistently.
- **Epochs 11–15**: Model stabilized with high accuracy (~89%) and decreasing validation loss, suggesting effective learning without overfitting.
- **Epochs 16–20**: Final training accuracy peaked at **91.71%**, and validation accuracy reached **89.50%**. Loss values converged to **0.2905 (train)** and **0.3586 (val)**.

####  Key Observations

- **Strong generalization**: Validation accuracy consistently tracked training accuracy, with no major divergence.
-  **Loss convergence**: Both training and validation loss decreased smoothly, indicating stable optimization.
-  **Model maturity**: By epoch 20, the model shows high confidence and low error, suitable for deployment or further tuning.



In [17]:
# Make predictions on the test set
y_pred_probs = model.predict(X_test)
y_pred = y_pred_probs.argmax(axis=1)

125/125 ━━━━━━━━━━━━━━━━━━━━ 0s 2ms/step


In [18]:
#  Evaluate performance
print("\nTest Accuracy:", accuracy_score(y_test, y_pred))
print("\nClassification Report:\n", classification_report(y_test, y_pred, target_names=le.classes_))


Test Accuracy: 0.8935

Classification Report:
               precision    recall  f1-score   support

           A       0.95      0.93      0.94       158
           B       0.84      0.93      0.88       153
           C       0.87      0.94      0.90       147
           D       0.90      0.88      0.89       161
           E       0.80      0.85      0.83       154
           F       0.87      0.83      0.85       155
           G       0.86      0.83      0.84       155
           H       0.82      0.81      0.81       147
           I       0.93      0.85      0.89       151
           J       0.89      0.93      0.91       149
           K       0.84      0.91      0.87       148
           L       0.91      0.93      0.92       152
           M       0.96      0.91      0.93       158
           N       0.91      0.89      0.90       157
           O       0.88      0.87      0.88       150
           P       0.92      0.89      0.91       161
           Q       0.96      0.96

####  Overall Performance

- **Test Accuracy**: `0.8935` → The model correctly predicted ~89.35% of the test samples.
- **Macro Avg F1-Score**: `0.89` → Balanced performance across all classes, regardless of support.
- **Weighted Avg F1-Score**: `0.89` → Strong overall performance, accounting for class imbalance.

####  Class-wise Metrics Breakdown

| Class | Precision | Recall | F1-Score | Support |
|-------|-----------|--------|----------|---------|
| A     | 0.95      | 0.93   | 0.94     | 158     |
| B     | 0.84      | 0.93   | 0.88     | 153     |
| C     | 0.87      | 0.94   | 0.90     | 147     |
| D     | 0.90      | 0.88   | 0.89     | 161     |
| E     | 0.80      | 0.85   | 0.83     | 154     |
| F     | 0.87      | 0.83   | 0.85     | 155     |
| G     | 0.86      | 0.83   | 0.84     | 155     |
| H     | 0.82      | 0.81   | 0.81     | 147     |
| I     | 0.93      | 0.85   | 0.89     | 151     |
| J     | 0.89      | 0.93   | 0.91     | 149     |
| K     | 0.84      | 0.91   | 0.87     | 148     |
| L     | 0.91      | 0.93   | 0.92     | 152     |
| M     | 0.96      | 0.91   | 0.93     | 158     |
| N     | 0.91      | 0.89   | 0.90     | 157     |
| O     | 0.88      | 0.87   | 0.88     | 150     |
| P     | 0.92      | 0.89   | 0.91     | 161     |
| Q     | 0.96      | 0.96   | 0.96     | 157     |
| R     | 0.84      | 0.81   | 0.82     | 151     |
| S     | 0.82      | 0.87   | 0.85     | 150     |
| T     | 0.90      | 0.91   | 0.90     | 159     |
| U     | 0.93      | 0.96   | 0.94     | 163     |
| V     | 0.94      | 0.93   | 0.93     | 153     |
| W     | 0.93      | 0.97   | 0.95     | 150     |
| X     | 0.90      | 0.83   | 0.86     | 157     |
| Y     | 0.92      | 0.92   | 0.92     | 157     |
| Z     | 0.94      | 0.90   | 0.92     | 147     |

####  Interpretation Highlights

-  **Top-performing classes**: Q, W, M, A — consistently high precision, recall, and F1.
-  **Lower-performing classes**: E, H, R — slightly lower F1-scores (~0.81–0.83), suggesting room for improvement.
-  **Balanced recall**: Most classes have recall ≥ 0.85, indicating good sensitivity across the board.
-  **Precision vs Recall trade-offs**: Class B shows lower precision (0.84) but high recall (0.93), hinting at false positives.




##**TASK-4:HYPERPARAMETER TUNING**

In [43]:
pip install scikeras --upgrade

In [46]:
from scikeras.wrappers import KerasClassifier
from sklearn.model_selection import RandomizedSearchCV
from tensorflow.keras.layers import Dropout

In [47]:
from scikeras.wrappers import KerasClassifier

In [50]:

def create_model():
    model = Sequential()
    model.add(Dense(32, input_dim=X_train.shape[1], activation='relu'))
    model.add(Dense(len(np.unique(y)), activation='softmax'))
    model.compile(optimizer='adam', loss='sparse_categorical_crossentropy', metrics=['accuracy'])
    return model


In [51]:
# Wrap with scikeras
model = KerasClassifier(model=create_model, epochs=5, batch_size=32, verbose=0)

In [52]:
# Fit and evaluate
model.fit(X_train, y_train)
print("Validation accuracy:", model.score(X_test, y_test))

/usr/local/lib/python3.12/dist-packages/keras/src/layers/core/dense.py:93: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(activity_regularizer=activity_regularizer, **kwargs)


Validation accuracy: 0.8055


In [53]:
from sklearn.model_selection import RandomizedSearchCV

def create_model(neurons=64, activation='relu', optimizer='adam'):
    model = Sequential()
    model.add(Dense(neurons, input_dim=X_train.shape[1], activation=activation))
    model.add(Dense(len(np.unique(y)), activation='softmax'))
    model.compile(optimizer=optimizer, loss='sparse_categorical_crossentropy', metrics=['accuracy'])
    return model

model = KerasClassifier(model=create_model, verbose=0)


In [54]:
param_dist = {
    "model__neurons": [32, 64, 128],
    "model__activation": ["relu", "tanh"],
    "model__optimizer": ["adam", "sgd"],
    "batch_size": [16, 32],
    "epochs": [10, 20]
}

random_search = RandomizedSearchCV(model, param_distributions=param_dist, n_iter=3, cv=2, verbose=2, random_state=42)
random_search_result = random_search.fit(X_train, y_train)

print("Best Params:", random_search_result.best_params_)

AttributeError: 'super' object has no attribute '__sklearn_tags__'